papas/path.py

```diff
@@ -114,11 +126,78 @@ def distquad (time):
         minim_answer = opti.bracket(distquad, xa = -0.5e-14, xb = 0.5e-14)
         self.closest_t = minim_answer[1]
         vector_IP = self.point_at_time(minim_answer[1]) - vertex
-        self.signIP= vector_IP.Dot(self.p4.Vect().Unit())
-        self.IP = minim_answer[4]**(1.0/2)*sign(self.signIP)
-        #self.closest=TVector3(self.fx(t), self.fy(t), self.fz(t))
+        Pj = jet.p4().Vect().Unit()
+        signIP  = vector_IP.Dot(Pj)
+        self.IP = minim_answer[4]**(1.0/2)*sign(signIP)
+        
+        x,y,z = self.coord_at_time(minim_answer[1])
+        self.IPx, self.IPy, self.IPz = (x-vertex.x()),  (y-vertex.y()), (z-vertex.z())
+       
+    def compute_theta_0(self):
+        '''Computes the square root of the variance, sigma, of the multiple
+        scattering angle due to matter interactions, using the formula in PDG
+        paper. The values for the beam pipe are set'''
+        # for the beam pipe,  X_0 = 1.72e-2, width = 1.5e-3
+        P = self.p4.Vect().Dot(self.udir)
+        PT= self.p4.Perp()
+        x = abs( epaisseur_beampipe * 1.0* P/PT)
+        X_0 = X_0_beampipe
+        # distance and radiation length, linked to the dectector properties
+        # and the particle's direction
+        self.theta_0 = 1.0*13.6e-3/(1.0*self.speed/constants.c*P)
+        self.theta_0 *= abs(self.charge)*(1.0*x/X_0)**(1.0/2)*(1+0.038*math.log(1.0*x/X_0))
+        self.xX_0 = 1.0*x/X_0
+    
+    def compute_IP_signif(self, IP, theta_0, scat_point):
+        # ! are we sure sigma_IP_due_IP_algo_precise isnt overestimated ?
+        delta_t = 1e-15
+        delta_s = delta_t * self.speed *1.0
+        sigma_s = delta_s
+        sigma_IP_due_IP_algo_precise = IP*1.0/(math.cos(math.atan(sigma_s/IP)))-IP
+        
+        if theta_0 == None or scat_point == None:
+            self.IP_signif = IP*1.0/sigma_IP_due_IP_algo_precise
+        else :        
+            phi_t_scat = self.phi( scat_point.X(), scat_point.Y())
+            t_scat = self.time_at_phi(phi_t_scat)
+            fly_distance = self.speed * 1.0 * t_scat
+            # for the IP significance : estimation 
+            sigma_IP_due_scattering = fly_distance*math.tan((2)**0.5*theta_0)
+            
+            sigma_IP_tot = ( sigma_IP_due_IP_algo_precise**2 + sigma_IP_due_scattering**2 )**0.5
+            
+            self.IP_signif = IP*1.0/sigma_IP_tot
+        
+    
+    def compute_new_dir(self):
+        '''Computes the new p4 vector due to multiple scattering while
```

Comment:
In fact, this method is useless : the job is done with multiple_scattering.py, so I removed it.